File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.py
from collections import Counter
from itertools import combinations
def solution(orders, course):
    answer = []
    # 최소 2명 이상의 손님으로부터 주문된 단품 메뉴 조합에 대해서만 후보로 포함
    # 각 손님은 2개이상 주문
    
    for c in course :
        candidates = []
        
        for order in orders: 
            order = sorted(order)
            
            for comb in combinations (order, c) :
                candidates.append(''.join(comb))

        counter = Counter(candidates)
        if not counter :
            continue

        max_count = max(counter.values())
        
        for menu,value in counter.items() :
            if max_count >=2 :
                if value == max_count :
                    answer.append(menu)
        
            
                
            
        
    
    return sorted(answer)

File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.java
import java.util.*;

class Solution {
    static Map<String, Integer> map;
    
    public String[] solution(String[] orders, int[] course) {
     
        ArrayList<String> answer = new ArrayList<>();
        
        for (int c: course) {
            map = new HashMap<>();
            
            
            for (String order : orders) {
                char [] ch_arr = order.toCharArray();
                Arrays.sort(ch_arr);
                
                combination(ch_arr,0,c,new StringBuilder());
            }
            
            int max_cnt = 0;
            
            for (int count : map.values()) {
                max_cnt = Math.max(max_cnt, count);
            }
            
            if (max_cnt <2) {
                continue;
            }
            
            
            for (String key: map.keySet()) {
                if(map.get(key)== max_cnt) {
                    answer.add(key);
                }
            }
            
            
        }
        
        Collections.sort(answer);
        return answer.toArray(new String[answer.size()]);
    }
    
    static void combination(char[] arr, int start, int targetLength, StringBuilder sb) {
        if(sb.length() == targetLength) {
            String menu = sb.toString();
            map.put(menu, map.getOrDefault(menu,0) +1);
            return;
        }
        
        for (int i =start; i <arr.length; i++) {
            sb.append(arr[i]);
            combination(arr,i+1,targetLength, sb) ;
            sb.deleteCharAt(sb.length()-1);
        }
    }
}